spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/dsl/GraphGeneratorVisitor.java

```diff
@@ -0,0 +1,651 @@
+					Node endNode = findOrMakeNode("$END");
+					addLink(new Link(currentTaskAppId, endNode.id, transition.getStateNameInDSLForm()));
+				}
+			} else {
+				Node n = existingNodesToReuse.get(transition.getTargetApp());
+				boolean isCreated = false;
+				if (n == null) {
+					String nextId = nextId();
+					n = new Node(nextId, transition.getTargetApp());
+					existingNodesToReuse.put(transition.getTargetApp(), n);
+					addNode(n);
+					isCreated = true;
+				}
+				addLink(new Link(currentTaskAppId, n.id, transition.getStatusToCheckInDSLForm()));
+				if (isCreated) {
+					if (currentContext().isFlow) {
+						currentContext().addOtherExit(n.id);
+					} else {
+						currentContext().addDanglingNodes(false, n.id);
+					}
+				}
+			}
+		} else {
+			// Check if it is a transition to something labeled earlier in this flow
+			Node existingNode = currentContext().getNodeLabeled(transition.getTargetLabel());
+			if (existingNode != null) {
+				addLink(new Link(currentTaskAppId, existingNode.id, transition.getStatusToCheckInDSLForm()));
+			}
+			else {
+				// Record a new transition attempted
+				currentContext().addTransitionTarget(currentTaskAppId, transition.getStatusToCheckInDSLForm(),
+						transition.getTargetLabel());
+			}
+		}
+	}
+
+	public GraphGeneratorVisitor.Context currentContext() {
+		return contexts.peek();
+	}
+
+	public GraphGeneratorVisitor.Context parentContext() {
+		if (contexts.size() < 2) {
+			return null;
+		} else {
+			return contexts.get(contexts.size() - 2);
+		}
+	}
+	
+	public Sequence currentSequence() {
+		return sequences.get(currentSequence);
+	}
+
+	@Override
+	public void visit(TaskApp taskApp) {
+		String nextId = nextId();
+		currentTaskAppId = nextId;
+		Node node = new Node(nextId, taskApp.getName());
+		addNode(node);
+		if (taskApp.hasLabel()) {
+			node.setLabel(taskApp.getLabelString());
+			currentContext().addNodeWithLabel(taskApp.getLabelString(),node);
+		}
+		if (currentContext().isFlow) {
+			// Are there any outstanding transitions that need to be connected
+			// to this?
+			if (taskApp.hasLabel()) {
+				for (Iterator<Context.TransitionTarget> iterator = currentContext().getTransitionTargets()
+						.iterator(); iterator.hasNext();) {
+					Context.TransitionTarget tt = iterator.next();
+					if (tt.label.equals(taskApp.getLabelString())) {
+						// Target found!
+						addLink(new Link(tt.nodeId, nextId, tt.onState));
+						iterator.remove();
+					}
+				}
+			}
+			List<String> danglingNodes = currentContext().getDanglingNodes();
+			if (danglingNodes.size() == 0) {
+				// first entry
+				addLink(new Link(currentContext().startNodeId, nextId));
+				currentContext().addDanglingNodes(true, nextId);
+			} else {
+				for (int i = 0; i < danglingNodes.size(); i++) {
+					addLink(new Link(danglingNodes.get(i), nextId));
+				}
+				currentContext().addDanglingNodes(true, nextId);
+			}
+			currentContext().clearDanglingNodes();
```

Comment:
I could be wrong but (which won't be the first time ;-)) But why are we clearing danglingNodes and re adding them if we are already clearing them and adding them in the if/else above.